Generalize the way properties of a contact constraints are computed, allow for custom implementations

peci1 opened the pull request:
This PR adds the possibility to adjust properties of contact constraints (joints) before they are added to the simulation loop.

It resolves a TODO from ContactConstraint.cpp that was reading:

> TODO(JS): Consider providing various ways of the combined friction or allowing to override this method by a custom method

The contact surface structure was inspired by the existing contact constraint class members and by http://ode.org/wiki/index.php?title=Manual#Contact .

This PR also adds support for specifying friction-unrelated contact velocity (`dContactMotion1` etc. in ODE words). This is needed for simulation of tracks or conveyor belts. It basically specifies that the velocity LCP tries to achieve should not be zero-up-to-friction-coefficients, but the specified velocity (very similar to how bouncing works). This velocity target can be set using the custom contact surface handler this PR adds.

Adding the custom contact surface handler also allowed to separate the odd-placed force-dependent slip settings from ConstraintSolver to the handler class, which looks to me like a desired separation of concerns.

This is the first step towards adding proper tracked vehicle support in SubT simulator, as requested here: https://github.com/osrf/subt/issues/365 . I verified viability of the proposed API by hacking it into ign-physics and using a custom handler from there. In the upcoming days, I'll try to come with a proper integration into ign-physics.

I did not take much care about keeping ABI compatibility. API compatibility should remain untouched regarding public members of classes. There are some new classes and I moved a few static protected members of ContactConstraint to DefaultContactSurfaceHandler. I also moved a few .cpp-only constants from ContactConstraint.cpp to ContactSurface.hpp.

The example API usage in my quick hack into ign-physics is like this:

```
class tmp : public dart::constraint::ContactSurfaceHandler
{
public:

  dart::constraint::ContactSurfaceParams createParams(const dart::collision::Contact& contact, const size_t numContactsOnCollisionObject) const override
  {
    auto params = this->mParent->createParams(contact);
    params.mFirstFrictionalDirection = Eigen::Vector3d::UnitY();
    params.mContactSurfaceMotionVelocity = Eigen::Vector3d::UnitY();
    return params;
  }
};

auto tmpPtr = std::make_shared<tmp>();

...

world->getConstraintSolver()->setContactSurfaceHandler(tmpPtr);
```


https://user-images.githubusercontent.com/182533/113764677-92ef4e80-971b-11eb-98e1-2c7fe3f3a7c6.mp4

diegoferigo commented:
Great work @peci1! I postponed the same task for too long (see https://github.com/ignition-forks/dart/issues/14), thanks for providing a way to expose these settings. If it's not too demanding, would it be possible exposing also the CFM parameter as discussed in https://github.com/ignition-forks/dart/issues/14#issuecomment-720633043? Otherwise, I can take care of it if this PR gets accepted.

xref: https://github.com/ignitionrobotics/ign-physics/issues/82, https://github.com/osrf/sdformat/issues/31, https://github.com/osrf/sdformat/issues/508, https://github.com/ignitionrobotics/ign-physics/pull/139.

peci1 commented:
Hi @diegoferigo. Quickly skimming through the code, it seemed to me that support for CFM setting would need more ABI breaking changes. I would like to keep this PR as "slim" as possible regarding ABI breaks (although there already are some). So I suggest you create another PR based on this one and you'll see whether the implementation provided by this PR is sufficient for the CFM case too. If not, we can make some adjustments so that it is still useful for you.

Just thinking about where to get the CFM values. One place is definitely the ODE SDF tag, which is kind of a bad practice but used for many other physics properties, so until osrf/sdformat#31 is resolved, it has to be done like that. Another way how CFM could be injected is via custom callbacks as is intended with the contact surface motion. You can choose whether you'd like to support this way of setting it, too, or not.

diegoferigo commented:
> Hi @diegoferigo. Quickly skimming through the code, it seemed to me that support for CFM setting would need more ABI breaking changes. I would like to keep this PR as "slim" as possible regarding ABI breaks (although there already are some). So I suggest you create another PR based on this one and you'll see whether the implementation provided by this PR is sufficient for the CFM case too. If not, we can make some adjustments so that it is still useful for you.

Totally fine with me. Exposing the CFM is one of my desiderata on the long term, and this can be done in a new PR. Since you started this task, I'm definitely willing to integrate with it. Once ABI are broken, let's try to merge as many changes as possible.

As an independent remark, I noticed a new [6.10 upstream release](https://github.com/dartsim/dart/releases/tag/v6.10.0). I'm not sure how it plays with the custom features implemented here and [optionally enabled in ign-physics](https://github.com/ignitionrobotics/ign-physics/search?q=DART_VERSION_AT_LEAST). Increasing the versioning of this fork might be required, and it could address ABI problems. The entire picture is not crystal clear to my eyes, maybe the developers can comment about it.

> Just thinking about where to get the CFM values. One place is definitely the ODE SDF tag, which is kind of a bad practice but used for many other physics properties, so until [osrf/sdformat#31](https://github.com/osrf/sdformat/issues/31) is resolved, it has to be done like that. Another way how CFM could be injected is via custom callbacks as is intended with the contact surface motion. You can choose whether you'd like to support this way of setting it, too, or not.

Using ODE tags was my original idea, waiting https://github.com/osrf/sdformat/issues/31. Good to know that it could be also implemented with custom callback, I'll keep this in mind.

peci1 commented:
@mjcarroll @scpeters if proper simulation of tracked vehicles should become a part of SubT simulator on time to be used in the challenge, I think we need to start discussing this PR as early as possible (it makes some ABI/API breaks and I'm not sure if that's okay, or if I should seek for some compatibility-retaining solutions). I'm sorry for putting pressure on you, but time is running.

peci1 commented:
Friendly ping to maintainers.

peci1 commented:
Videos of ign-gazebo implementation (they only work in Chrome, I don't know why).

https://user-images.githubusercontent.com/182533/122688768-75ac2580-d21e-11eb-9947-67844b548381.mp4


https://user-images.githubusercontent.com/182533/122688763-71800800-d21e-11eb-80b0-0fc894cdd7de.mp4

https://user-images.githubusercontent.com/182533/122688779-8bb9e600-d21e-11eb-8e36-b23a035fb8b5.mp4


https://user-images.githubusercontent.com/182533/122688788-9c6a5c00-d21e-11eb-9e18-60d625bc8761.mp4

peci1 commented:
This PR allows https://github.com/ignitionrobotics/ign-physics/pull/267, https://github.com/ignitionrobotics/ign-gazebo/pull/869 and https://github.com/osrf/subt/pull/958 .

chapulina reviewed (changes requested):
> it makes some ABI/API breaks and I'm not sure if that's okay, or if I should seek for some compatibility-retaining solutions

That would trigger a cascade of incompatibility issues across all stable Ignition versions. We already have a lot to handle with the various debs coming from DART's PPA and our fork, which vary according to Ubuntu version. I strongly recommend looking into compatibility-retaining solutions.

scpeters commented on dart/constraint/ContactConstraint.hpp:
```diff
@@ -132,19 +138,6 @@ class ContactConstraint : public ConstraintBase
-      const dynamics::ShapeNode* shapenode);
-  static double computeRestitutionCoefficient(
-      const dynamics::ShapeNode* shapeNode);
-
```
to avoid breaking API, can you keep these functions, even if they aren't used anymore? you could deprecate them instead

peci1 replied:
I added the functions back

scpeters commented on dart/constraint/ConstraintSolver.hpp:
```diff
@@ -259,6 +277,9 @@ class ConstraintSolver
   std::vector<ConstrainedGroup> mConstrainedGroups;
+
+  /// Factory for ContactSurfaceParams for each contact
+  ContactSurfaceHandlerPtr mContactSurfaceHandler;
```
I think this has an ABI impact

nkoenig replied:
Do you think there is an alternative approach that wouldn't break ABI?

scpeters replied:
I don't see any obvious solutions

peci1 replied:
I'll sketch the standard ABI-retain hack with a global `std::map` and mutex.

scpeters commented on dart/constraint/ContactConstraint.hpp:
```diff
@@ -210,6 +203,12 @@ class ContactConstraint : public ConstraintBase
+  /// x = vel. in direction of contact normal
+  /// y = vel. in first friction direction
+  /// z = vel. in second friction direction
+  Eigen::Vector3d mContactSurfaceMotionVelocity;
```
I think this has an ABI impact

peci1 replied:
resolved using global map and mutex

azeey commented on dart/constraint/ContactSurface.hpp:
```diff
@@ -0,0 +1,160 @@
+  Eigen::Vector3d mContactSurfaceMotionVelocity {DART_DEFAULT_CONTACT_SURFACE_MOTION_VELOCITY};
+};
+
+class ContactConstraint;
```
nit: Is this needed here?

peci1 replied:
Good catch, that was a remnant from the development phase... removed

azeey commented on dart/constraint/ContactSurface.hpp:
```diff
@@ -0,0 +1,160 @@
+  /// \param[in] parent Optional parent handler. It suggested that new handlers
+  ///                   should not overwrite the previous ones, but rather
+  ///                   "extend" them.
+  explicit ContactSurfaceHandler(ContactSurfaceHandlerPtr parent = nullptr);
```
Can you elaborate on the design of this class? I see a hierarchical system where you might have a chain of `ContactSurfaceHandler`s created by setting the `parent` here or using `setParent`. The class can also be inherited from to create a similar hierarchy. Is there a reason for having two ways of doing this?

peci1 replied:
The idea is simple - if this code ever gets upstreamed, it should allow several independent "clients" to register their handlers. One such "client" can be ign-physics, another could be e.g. a debugging snippet that reads or changes some contact properties. Inheritance would not help in such case. Maybe the wording in the comment is a bit misleading.

peci1 replied:
I tried to improve the wording.

azeey replied:
Ah, I see. I hadn't considered the use case with multiple clients. Thanks for the explanation.

peci1 commented:
I resolved (hopefully) all ABI breaking parts of this PR. Could you please re-review it?

chapulina reviewed (commented):
Thanks for addressing the ABI issue! It would be nice to add a comment to the hack so people know why it's there and can come back to fix it in a future (major?) version.

Removing my request for changes.

azeey commented on dart/constraint/ContactSurface.hpp:
```diff
@@ -0,0 +1,162 @@
+  /// y = vel. in first friction direction
+  /// z = vel. in second friction direction
+  Eigen::Vector3d mContactSurfaceMotionVelocity {DART_DEFAULT_CONTACT_SURFACE_MOTION_VELOCITY};
+};
```
If we're going through all this trouble to keep ABI compatibility, it would be nice to use the PIMPL idiom to make this extensible. Is that feasible?

peci1 replied:
Do you mean hiding the whole struct's contents in a PIMPL and adding public non-virtual getters/setters to the struct? I'm not sure what could be the performance impact of creating and destroying the unique pointers, as this structure is pretty short-lived and there can theoretically be thousands of them created every simulation step.

azeey replied:
Yes, that is what I meant. I share your concern on the performance impact. I know there is interest in adding more parameters (https://github.com/ignition-forks/dart/pull/22#issuecomment-817376968), but doing so would break ABI unless (a) we add those parameters now, (b) we provide a means for adding them later on without braking ABI

What about a hybrid approach where we add a pointer to extra parameters that is only allocated on demand. For now, the current list of parameters will be directly available in the struct, but new parameters will be added in a new struct hidden to end users and only made accessible via getters/setters.

azeey reviewed (commented):
Thanks for the great contribution. This looks pretty good. I have a couple of questions regarding the design of `ContactSurfaceHandler` and the `ContactSurfaceParams` struct.

peci1 commented:
> It would be nice to add a comment to the hack so people know why it's there and can come back to fix it in a future (major?) version.

Comments added.

azeey commented on dart/constraint/ConstraintSolver.cpp:
```diff
@@ -542,27 +574,28 @@ void ConstraintSolver::updateConstraints()
-        contactConstraint->getSecondarySlipCompliance() * numContacts);
+    ContactConstraintPtr contactConstraint;
+    {
+      std::lock_guard<std::mutex> lock(gContactSurfaceHandlersMutex);
```
Can the mutex lock and fetching of the constraint surface handler be moved outside of the for loop? I think it would help performance.

peci1 replied:
done

azeey commented on dart/constraint/ConstraintSolver.hpp:
```diff
@@ -184,6 +184,24 @@ class ConstraintSolver
+  /// Set the handler used for computing contact surface parameters based on
+  /// the contact properties of the two colliding bodies. This function
+  /// automatically sets the previous handler as parent of the given handler.
+  void setContactSurfaceHandler(ContactSurfaceHandlerPtr handler);
```
Since this is really adding a new handler, how about calling it `addContactSurfaceHandler`?

peci1 replied:
done

azeey commented on dart/constraint/ConstraintSolver.hpp:
```diff
@@ -184,6 +184,24 @@ class ConstraintSolver
 
+  /// Get the handler used for computing contact surface parameters based on
+  /// the contact properties of the two colliding bodies.
+  ContactSurfaceHandlerPtr getContactSurfaceHandler() const;
```
Since this is getting the last handler in the chain of handlers, how about calling it `getLastContactSurfaceHandler`?

peci1 replied:
done

peci1 commented:
@azeey I addressed all your comments. Now I'm leaving for 5-6 days, so if you want some minor changes, feel free to do them yourself in order to speed up acceptance of this PR.

azeey commented on dart/constraint/ConstraintSolver.cpp:
```diff
@@ -542,31 +576,31 @@ void ConstraintSolver::updateConstraints()
+      if (it != contactPairMap.end())
+        numContacts = it->second;
+
+      auto contactConstraint = gContactSurfaceHandlers[this]->createConstraint(
```
nit: The search for the handler, i.e. `gContactSurfaceHandlers[this]` could be moved outside of the loop.

azeey replied:
Will address this in a followup PR.

azeey reviewed (approved):
Thanks for the great contribution and for iterating through the review process.
